File: src/app/keymap.ts
import { useEditor } from '@/features/editor/openFiles'
import { panelsOf, prune } from '@/lib/layout'
import { hasPrimaryModifier, isMac } from '@/lib/platform'
import type { ActionBinding, PanelId } from '@/types/beacon'
import { focusPanel, panelAfter, usePanelFocus } from './panelFocus'
import { selectHidden, selectLayout, useBeacon } from './store'

/**
 * What each bindable action does.
 *
 * The catalogue of actions and their defaults lives in the backend, which is
 * what lets conflict checking have one source of truth. What they *do* lives
 * here, keyed by the same ids. An action in one and not the other is a bug, and
 * `missingHandlers` is how it gets noticed rather than silently doing nothing.
 */
const HANDLERS: Record<string, () => void> = {
  'palette.open': () => useBeacon.getState().setOverlay('palette'),
  'quickOpen.open': () => useBeacon.getState().setOverlay('quickOpen'),
  'settings.open': () => useBeacon.getState().setOverlay('settings'),

  'panel.toggle.files': () => void useBeacon.getState().togglePanel('files'),
  'panel.toggle.git': () => void useBeacon.getState().togglePanel('git'),
  'panel.toggle.editor': () => void useBeacon.getState().togglePanel('editor'),
  'panel.toggle.terminal': () => void useBeacon.getState().togglePanel('terminal'),

  'panel.fullscreen': () => {
    const store = useBeacon.getState()
    store.toggleFullscreen(store.fullscreenPanel ?? 'claude')
  },

  'panel.focusNext': () => shiftFocus(1),
  'panel.focusPrevious': () => shiftFocus(-1),

  // Bound at the application level, not only inside CodeMirror: the file with
  // unsaved changes is the one you were just editing, and Cmd+S doing nothing
  // because focus had moved to a tab or the tree is how work gets lost.
  'editor.save': () => {
    const store = useBeacon.getState()
    const workspace = store.snapshot?.activeWorkspace
    const project = activeProjectId(store)
    if (!workspace || !project) return

    const editor = useEditor.getState()
    const open = editor.byProject[project] ?? []
    const active = open.find((file) => file.path === editor.active[project]) ?? open.at(-1)
    if (active) void editor.save(workspace, project, active.path)
  },

  'session.restartClaude': () => {
    const store = useBeacon.getState()
    const project = activeProjectId(store)
    if (project) void store.restartSession(project, 'claude')
  },

  'project.next': () => shiftProject(1),
  'project.previous': () => shiftProject(-1),
}

/** Human-facing names, kept here because they are interface text. */
export const ACTION_TITLES: Record<string, string> = {
  'palette.open': 'Command palette',
  'quickOpen.open': 'Quick open',
  'settings.open': 'Settings',
  'panel.toggle.files': 'Toggle Files',
  'panel.toggle.git': 'Toggle Git',
  'panel.toggle.editor': 'Toggle the editor',
  'panel.toggle.terminal': 'Toggle the terminal',
  'panel.fullscreen': 'Fullscreen the focused panel',
  'panel.focusNext': 'Focus the next panel',
  'panel.focusPrevious': 'Focus the previous panel',
  'editor.save': 'Save the file',
  'session.restartClaude': 'Restart Claude',
  'project.next': 'Next project',
  'project.previous': 'Previous project',
}

/**
 * Moves the keyboard to the next panel along.
 *
 * The order is the layout's own, so it follows what you can see: a fullscreen
 * panel is the only one there is, and a hidden panel is not somewhere focus can
 * land.
 */
function shiftFocus(step: number): void {
  const store = useBeacon.getState()
  const layout = selectLayout(store)

  const order: PanelId[] = store.fullscreenPanel
    ? [store.fullscreenPanel]
    : layout
      ? panelsOf(prune(layout, selectHidden(store)) ?? layout)
      : []

  const next = panelAfter(order, usePanelFocus.getState().focused, step)
  if (next) focusPanel(next)
}

/** Actions the backend offers that nothing here implements. */
export function missingHandlers(bindings: ActionBinding[]): string[] {
  return bindings.map((b) => b.action).filter((action) => !(action in HANDLERS))
}

/**
 * The binding an event describes, in the same form the backend stores.
 *
 * Returns `null` for anything without the primary modifier, since every Beacon
 * shortcut has one — that is what stops them firing while you type.
 */
export function bindingOf(event: KeyboardEvent): string | null {
  if (!hasPrimaryModifier(event)) return null

  const key = event.key.toLowerCase()
  // A modifier on its own is not a shortcut.
  if (['shift', 'alt', 'control', 'meta', 'os'].includes(key)) return null

  const parts = ['mod']
  if (event.shiftKey) parts.push('shift')
  if (event.altKey) parts.push('alt')
  parts.push(key === ' ' ? 'space' : key)
  return parts.join('+')
}

/** Runs whatever is bound to this event, and says whether anything was. */
export function runBinding(event: KeyboardEvent, bindings: ActionBinding[]): boolean {
  const pressed = bindingOf(event)
  if (!pressed) return false

  const match = bindings.find((binding) => binding.binding === pressed)
  if (!match) return false

  const handler = HANDLERS[match.action]
  if (!handler) return false

  handler()
  return true
}

/** `mod+shift+p` as `⌘⇧P`, or `Ctrl+Shift+P`. */
export function describeBinding(binding: string): string {
  const parts = binding.split('+')
  const key = parts.at(-1) ?? ''
  const shift = parts.includes('shift')
  const alt = parts.includes('alt')

  const label = KEY_LABELS[key] ?? key.toUpperCase()

  return isMac()
    ? `${alt ? '⌥' : ''}${shift ? '⇧' : ''}⌘${label}`
    : `Ctrl+${shift ? 'Shift+' : ''}${alt ? 'Alt+' : ''}${label}`
}

const KEY_LABELS: Record<string, string> = {
  enter: '↩',
  space: '␣',
  escape: '⎋',
  arrowup: '↑',
  arrowdown: '↓',
  arrowleft: '←',
  arrowright: '→',
  ',': ',',
  '[': '[',
  ']': ']',
}

function activeProjectId(store: ReturnType<typeof useBeacon.getState>): string | undefined {
  const snapshot = store.snapshot
  const workspace = snapshot?.workspaces.find((w) => w.id === snapshot.activeWorkspace)
  if (!workspace) return undefined
  return snapshot?.activeProject[workspace.id] ?? workspace.projects[0]?.id
}

/** Moves to the next or previous tab, wrapping around. */
function shiftProject(step: number): void {
  const store = useBeacon.getState()
  const snapshot = store.snapshot
  const workspace = snapshot?.workspaces.find((w) => w.id === snapshot.activeWorkspace)
  const projects = workspace?.projects ?? []
  if (projects.length === 0) return

  const current = projects.findIndex((project) => project.id === activeProjectId(store))
  const next = (current + step + projects.length) % projects.length
  void store.selectProjectAt(next)
}

export { HANDLERS as BOUND_ACTIONS }

File: crates/beacon-core/src/keymap.rs
use std::collections::BTreeMap;

use serde::Serialize;

use crate::error::{CoreError, Result};

/// Everything that can be given a keyboard shortcut, and what it starts as.
///
/// Only stable actions belong here. The command palette also lists things like
/// "switch to this particular project", whose identity changes with the
/// workspace — those are discoverable but not bindable, because a binding has
/// to mean the same thing next week.
///
/// The table lives in the backend so that conflict checking has one source of
/// truth. What each action *does* lives in the frontend, keyed by these ids.
pub const ACTIONS: &[(&str, &str)] = &[
    ("palette.open", "mod+k"),
    ("quickOpen.open", "mod+p"),
    ("settings.open", "mod+,"),
    ("panel.toggle.files", "mod+e"),
    ("panel.toggle.git", "mod+g"),
    ("panel.toggle.editor", "mod+o"),
    ("panel.toggle.terminal", "mod+j"),
    ("panel.fullscreen", "mod+enter"),
    // Arrows rather than letters, because Option rewrites what a letter key
    // reports on macOS and the binding would arrive as a character nobody
    // pressed.
    ("panel.focusNext", "mod+alt+arrowright"),
    ("panel.focusPrevious", "mod+alt+arrowleft"),
    ("editor.save", "mod+s"),
    ("session.restartClaude", "mod+shift+r"),
    ("project.next", "mod+shift+]"),
    ("project.previous", "mod+shift+["),
];

/// Bindings Beacon refuses to take over.
///
/// These belong to the terminal, and a session is the point of the window.
/// Rebinding copy or paste would break them inside Claude and inside every
/// shell, and the cause would be almost impossible to guess at from the
/// symptom.
const RESERVED: &[&str] = &["mod+c", "mod+v", "mod+x", "mod+a", "mod+z"];

/// One action as the settings screen needs it.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ActionBinding {
    pub action: String,
    /// What it is bound to now.
    pub binding: String,
    /// What it would be with nothing configured.
    pub default_binding: String,
}

pub fn is_known_action(action: &str) -> bool {
    ACTIONS.iter().any(|(id, _)| *id == action)
}

pub fn default_binding(action: &str) -> Option<&'static str> {
    ACTIONS
        .iter()
        .find(|(id, _)| *id == action)
        .map(|(_, binding)| *binding)
}

/// Puts a binding into the one form everything else compares against.
///
/// Accepts what a person would type — `Cmd+Shift+R`, `ctrl+shift+r` — and
/// answers in lower case with modifiers in a fixed order, so two spellings of
/// the same shortcut can never both be stored.
pub fn normalize(binding: &str) -> Result<String> {
    let mut modifier = false;
    let mut shift = false;
    let mut alt = false;
    let mut key: Option<String> = None;

    for part in binding.split('+') {
        let part = part.trim().to_ascii_lowercase();
        if part.is_empty() {
            continue;
        }

        match part.as_str() {
            // Beacon's shortcuts are written against "the primary modifier" so
            // one table is correct on macOS and Linux; all of these mean it.
            "mod" | "cmd" | "command" | "ctrl" | "control" | "meta" | "super" => modifier = true,
            "shift" => shift = true,
            "alt" | "option" | "opt" => alt = true,
            other => {
                if key.is_some() {
                    return Err(CoreError::invalid(
                        "a shortcut can only have one key besides its modifiers",
                    ));
                }
                key = Some(other.to_string());
            }
        }
    }

    let key = key.ok_or_else(|| CoreError::invalid("a shortcut needs a key"))?;
    if !modifier {
        return Err(CoreError::invalid(
            "shortcuts must include the primary modifier, so they cannot collide with typing",
        ));
    }

    let mut parts = vec!["mod"];
    if shift {
        parts.push("shift");
    }
    if alt {
        parts.push("alt");
    }
    let normalized = format!("{}+{key}", parts.join("+"));

    if RESERVED.contains(&normalized.as_str()) {
        return Err(CoreError::invalid(format!(
            "{normalized} belongs to the terminal; rebinding it would break it inside every session"
        )));
    }

    Ok(normalized)
}

/// Every action with the binding it currently has.
pub fn resolve(overrides: &BTreeMap<String, String>) -> Vec<ActionBinding> {
    ACTIONS
        .iter()
        .map(|(action, fallback)| ActionBinding {
            action: (*action).to_string(),
            binding: overrides
                .get(*action)
                .cloned()
                .unwrap_or_else(|| (*fallback).to_string()),
            default_binding: (*fallback).to_string(),
        })
        .collect()
}

/// Which other action already answers to this binding, if any.
pub fn conflicting_action(
    overrides: &BTreeMap<String, String>,
    binding: &str,
    excluding: &str,
) -> Option<String> {
    resolve(overrides)
        .into_iter()
        .find(|entry| entry.binding == binding && entry.action != excluding)
        .map(|entry| entry.action)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn every_default_is_a_valid_binding() {
        for (action, binding) in ACTIONS {
            normalize(binding)
                .unwrap_or_else(|err| panic!("{action}'s default {binding} is invalid: {err}"));
        }
    }

    #[test]
    fn no_two_actions_start_out_sharing_a_binding() {
        let mut seen: Vec<&str> = ACTIONS.iter().map(|(_, binding)| *binding).collect();
        let before = seen.len();
        seen.sort_unstable();
        seen.dedup();
        assert_eq!(before, seen.len(), "two actions share a default shortcut");
    }

    #[test]
    fn the_same_shortcut_spelled_differently_normalizes_the_same_way() {
        let forms = ["Cmd+Shift+R", "ctrl+shift+r", "  meta + SHIFT + r "];
        for form in forms {
            assert_eq!(normalize(form).unwrap(), "mod+shift+r", "for {form:?}");
        }
    }

    #[test]
    fn modifiers_come_back_in_a_fixed_order() {
        assert_eq!(normalize("mod+alt+shift+k").unwrap(), "mod+shift+alt+k");
    }

    #[test]
    fn a_shortcut_without_the_primary_modifier_is_refused() {
        // Otherwise it would fire while typing into the terminal.
        assert!(normalize("shift+k").is_err());
        assert!(normalize("k").is_err());
    }

    #[test]
    fn a_shortcut_with_no_key_is_refused() {
        assert!(normalize("mod+shift").is_err());
        assert!(normalize("").is_err());
    }

    #[test]
    fn two_keys_are_refused() {
        assert!(normalize("mod+k+p").is_err());
    }

    #[test]
    fn the_terminals_own_shortcuts_are_refused() {
        for reserved in ["mod+c", "Cmd+V", "ctrl+x"] {
            let error = normalize(reserved).unwrap_err().to_string();
            assert!(error.contains("terminal"), "should explain why: {error}");
        }
    }

    #[test]
    fn resolving_prefers_an_override_and_still_reports_the_default() {
        let mut overrides = BTreeMap::new();
        overrides.insert("palette.open".to_string(), "mod+shift+p".to_string());

        let resolved = resolve(&overrides);
        let palette = resolved
            .iter()
            .find(|e| e.action == "palette.open")
            .unwrap();
        assert_eq!(palette.binding, "mod+shift+p");
        assert_eq!(palette.default_binding, "mod+k");

        // Untouched actions keep theirs.
        let quick = resolved
            .iter()
            .find(|e| e.action == "quickOpen.open")
            .unwrap();
        assert_eq!(quick.binding, "mod+p");
    }

    #[test]
    fn a_binding_already_in_use_is_reported_with_who_has_it() {
        let overrides = BTreeMap::new();
        assert_eq!(
            conflicting_action(&overrides, "mod+p", "palette.open").as_deref(),
            Some("quickOpen.open")
        );
        // An action does not conflict with itself.
        assert_eq!(
            conflicting_action(&overrides, "mod+k", "palette.open"),
            None
        );
    }

    #[test]
    fn a_binding_freed_by_an_override_stops_conflicting() {
        let mut overrides = BTreeMap::new();
        overrides.insert("quickOpen.open".to_string(), "mod+shift+p".to_string());
        assert_eq!(
            conflicting_action(&overrides, "mod+p", "palette.open"),
            None
        );
    }
}
